posthog/clickhouse/migrations/0064_materialize_elements_chain.py

```diff
@@ -0,0 +1,37 @@
+from infi.clickhouse_orm import migrations
+
+from posthog.clickhouse.client.connection import ch_pool
+from posthog.clickhouse.client.migration_tools import run_sql_with_exceptions
+from posthog.models.event.sql import (
+    EVENTS_TABLE_JSON_MV_SQL,
+    KAFKA_EVENTS_TABLE_JSON_SQL,
+)
+from posthog.settings import CLICKHOUSE_CLUSTER
+
+
+# KafkaEngine doesn't support DEFAULT/MATERIALIZED/EPHEMERAL expressions for columns, so we're
+# leaning on the default of 0 to be "full" (as desired)
+ADD_COLUMNS_BASE_SQL = """
+ALTER TABLE {table} ON CLUSTER {cluster}
+ADD COLUMN IF NOT EXISTS elements_chain_mat_href String MATERIALIZED extract(elements_chain, '(?::|\")href="(.*?)"'),
+ADD COLUMN IF NOT EXISTS elements_chain_mat_texts Array(String) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?::|\")text="(.*?)"')),
+ADD COLUMN IF NOT EXISTS elements_chain_mat_ids Array(String) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?::|\")id="(.*?)"')),
+ADD COLUMN IF NOT EXISTS elements_chain_mat_elements Array(Enum('a', 'button', 'form', 'input', 'select', 'textarea', 'label')) MATERIALIZED arrayDistinct(extractAll(elements_chain, '(?:^|;)(a|button|form|input|select|textarea|label)(?:\\.|$|:)'))
+"""
+
+
+def add_columns_to_required_tables(_):
+    with ch_pool.get_client() as client:
+        client.execute(ADD_COLUMNS_BASE_SQL.format(table="events", cluster=CLICKHOUSE_CLUSTER))
+
+        client.execute(ADD_COLUMNS_BASE_SQL.format(table="writable_events", cluster=CLICKHOUSE_CLUSTER))
```

Comment:
Even though you technically _can_ create `MATERIALIZED` columns on a distributed table like this now, this goes against how we do materialization. I'd take a look at `events` and `sharded_events` to see the difference, but the gist is you only create the materialized column on the `sharded` (actual MergeTree) table and then read through through that on the `events` distributed table. **You _do not_ want to put _any_ materialized columns or references to materialized columns on the writable table. This can break ingestion.**

You can see that on the `writable_events` table as it is on prod right now:
```sql
CREATE TABLE posthog.writable_events
(
    `uuid` UUID,
    `event` String,
    `properties` String,
    `timestamp` DateTime64(6, 'UTC'),
    `team_id` Int64,
    `distinct_id` String,
    `elements_hash` String,
    `created_at` DateTime64(6, 'UTC'),
    `_timestamp` DateTime,
    `_offset` UInt64,
    `elements_chain` String,
    `person_id` UUID,
    `person_properties` String,
    `group0_properties` String,
    `group1_properties` String,
    `group2_properties` String,
    `group3_properties` String,
    `group4_properties` String,
    `person_created_at` DateTime64(3),
    `group0_created_at` DateTime64(3),
    `group1_created_at` DateTime64(3),
    `group2_created_at` DateTime64(3),
    `group3_created_at` DateTime64(3),
    `group4_created_at` DateTime64(3),
    `inserted_at` Nullable(DateTime64(6, 'UTC')) DEFAULT NULL,
    `person_mode` Enum8('full' = 0, 'propertyless' = 1, 'force_upgrade' = 2)
)
ENGINE = Distributed('posthog', 'posthog', 'sharded_events', sipHash64(distinct_id))
```

`writable_events` represents the schema coming in from Kafka. All materialization occurs on the `MergeTree` table.

Basically you need to do two things here:
- Create the `MATERIALIZED` column on `sharded_events`
- Add a reference to it from `events`...something like 
```
`mat_ec_<column_name>` String COMMENT 'column_materializer_ec::<column_name>',
```

I used something similar here to how we do mat cols currently for properties, just added an `_ec` for elements chain. 

You don't need to change `writable_events`.